cloudinfo/cbr_test.go

```diff
@@ -0,0 +1,181 @@
+    "contexts": [
+        {
+            "attributes": [
+                {
+                    "name": "endpointType",
+                    "value": "private"
+                },
+                {
+                    "name": "networkZoneId",
+                    "value": "366d56825bd272e7e7ffe0299e74f22b"
+                }
+            ]
+        }
+    ],
+    "created_at": "2023-02-27T17:11:51.000Z",
+    "created_by_id": "IBMid-270006EYUC",
+    "crn": "crn:v1:bluemix:public:context-based-restrictions:global:a/abac0df06b644a9cabc6e44f55b3880e::rule:ca1c2bb48b40ed7c595a6ff3ed71a15c",
+    "description": "test-postgres-postgres access only from vpc",
+    "enforcement_mode": "enabled",
+    "href": "https://cbr.cloud.ibm.com/v1/rules/ca1c2bb48b40ed7c595a6ff3ed71a15c",
+    "id": "ca1c2bb48b40ed7c595a6ff3ed71a15c",
+    "last_modified_at": "2023-02-27T17:11:51.000Z",
+    "last_modified_by_id": "IBMid-270006EYUC",
+    "operations": {
+        "api_types": [
+            {
+                "api_type_id": "crn:v1:bluemix:public:context-based-restrictions::::api-type:data-plane"
+            }
+        ]
+    },
+    "resources": [
+        {
+            "attributes": [
+                {
+                    "name": "accountId",
+                    "operator": "stringEquals",
+                    "value": "abac0df06b644a9cabc6e44f55b3880e"
+                },
+                {
+                    "name": "serviceInstance",
+                    "operator": "stringEquals",
+                    "value": "c2ce344a-752c-4d9b-9374-75a0b664aecf"
+                },
+                {
+                    "name": "serviceName",
+                    "operator": "stringEquals",
+                    "value": "databases-for-postgresql"
+                }
+            ]
+        }
+    ]
+}`
+const zoneJson = `{
+  "account_id": "abac0df06b644a9cabc6e44f55b3880e",
+  "address_count": 1,
+  "addresses": [
+    {
+      "type": "vpc",
+      "value": "crn:v1:bluemix:public:is:us-south:a/abac0df06b644a9cabc6e44f55b3880e::vpc:r006-746aefab-fb68-4aea-bd9f-6ad069a72288"
+    }
+  ],
+  "created_at": "2023-02-27T16:22:43.000Z",
+  "created_by_id": "IBMid-270006EYUC",
+  "crn": "crn:v1:bluemix:public:context-based-restrictions:global:a/abac0df06b644a9cabc6e44f55b3880e::zone:366d56825bd272e7e7ffe0299e74f22b",
+  "description": "CBR Network zone representing VPC",
+  "excluded": [],
+  "excluded_count": 0,
+  "href": "https://cbr.cloud.ibm.com/v1/zones/366d56825bd272e7e7ffe0299e74f22b",
+  "id": "366d56825bd272e7e7ffe0299e74f22b",
+  "last_modified_at": "2023-02-27T16:22:43.000Z",
+  "last_modified_by_id": "IBMid-270006EYUC",
+  "name": "test-postgres-VPC-network-zone"
+}`
+
+// TestGetCBRRuleByID tests the GetCBRRuleByID function with a valid rule ID
+func TestGetCBRRuleByID(t *testing.T) {
+
+	ruleID := "ca1c2bb48b40ed7c595a6ff3ed71a15c"
+
+	var mockRule = contextbasedrestrictionsv1.Rule{}
+
+	err := json.Unmarshal([]byte(ruleJson), &mockRule)
```

Comment:
if this is giving you too much trouble, you might not really need the complex json, since your mocked functions don't actually need or do anything with it. You could just produce a much simpler object in your mock function that gets returned just so you can verify.

I think the most important part of the tests are that your actual service functions return the appropriate error or non-error conditions based on your mock settings.

Reply:
For example, since the struct for Rule and Zone are defined, you can just set them up yourself. Like
```go
var mockRule = &contextbasedrestrictionsv1.Rule{
    ID: "is-just-fake-1234",
    Description: "doesnt really matter"
```
In your test you really only need ID set, golang won't care you didn't set the rest of them, and your mock functions don't rely on them. Your real functions are so basic you might not even need to check the ID of the output, what you are really after is if it is nil or not based on error conditions.